### PR TITLE
Add initial check enforcer functionality

### PR DESCRIPTION
**Pull request by benbp**
This adds a new implementation of check enforcer intended to be run within a github actions context. To see an example of it running for a repository, check out the [actions in my fork](https://github.com/benbp/azure-sdk-tools/actions).

## Design Goals and Current Problems:

- Reduce ongoing maintenance and knowledge costs of the current check enforcer. By using github actions we won't have to maintain a check enforcer service ourselves.
- Improve check enforcer scaling. While the current check enforcer app can process webhook events much more quickly than github actions (since it does not have to spin up a runner VM per event), it hits a few key limitations:

## Details

- Check enforcer maintenance
    - There is a lot of work that needs to be done to improve scaling problems with check enforcer and support rapid development and safe deployments, detailed in https://github.com/Azure/azure-sdk-tools/issues/3033. Moving to this github action seems is less work for more benefit.
- Check enforcer scaling issues
    - The github API has a secondary rate limit for parallel requests from a single github app identity.
    - check enforcer handles events for all of our repositories.
    - check enforcer operates as part of a check suite, and queries the status of all check runs in the suite every time an individual run completes. When a large changeset is put in a pull request or merged to main, it causes a build storm and therefore a check run storm. Check enforcer not only has to query all check runs for each of these, but github paginates the responses multiplying the high number of requests by around 50 in the worst cases.
    - The above and other behaviors cause big queue backups due to github throttling in extreme cases, which happen with relative frequency (every few weeks) and bring check enforcer to a halt for hours.
- Github action benefits to scaling
    - When github runs an action, it passes a generated token in the [workflow context](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow) that can be used for requests. From what I can tell, this token will not have the same "one request at a time" limitations as our current github app identity, but I need to do more load testing from the action to understand when we'll get rate limited.
    - Since actions are local to the repo, even if we hit throttling issues due to build storms (or queueing issues due to event storms), this will only affect a single repo rather than all repos.

## Design Decisions

There are a few key design decisions made with this change, that I am eager will spark discussion.

1. As an organization we are starting to recognize more areas in which github actions can provide value that is not available today via an azure pipelines solution. These primarily center around ease of onboarding, discovery and understanding, as well as much more flexibility around various repository events.
    - As we continue to add github actions, I think it is going to be optimal to provide common platform support from the engsys team. Some ideas I have here are having centralized action configs for each event type, to prevent more than one action being queued for the same event. We can handle all the event details, routing, status checking, logging, testing, etc. and reduce duplicate efforts by providing a framework that developers can build their actions in.
1. The tool is written in Go. This decision is influenced by a future vision for our usage of github actions as an organization.
    - A top-level design goal should be to reduce the overhead of running the tool as much as possible, as our usage of actions grows (especially if we move more of our internal pipeline workloads into actions).
    - This tool or similar tools are likely to grow in complexity, usage and coverage. Tasks that contribute to setup time could include tool downloading, dependency installation, runtime startup, JIT compiling, etc. Go has a very minimal overhead in this area vs. Python, C#, Powershell, etc.
    - Any approach we choose needs to be easy to pick up by different language teams. Currently we use powershell as a common denominator, but there are many benefits to using a static language with static binaries that powershell cannot offer. Additionally, enforcing powershell quality, conventions and testing standards is extremely difficult today. Go has a much more opinionated and strict approach to how programs are written, which will help with consistency and ease of understanding of programs written across the repositories.

## Implementation Details

The program is currently written to take a single argument which is a path to a payload file. The payload is expected to be some form of a github event. Currently, it supports two payloads: [check_suite](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#check_suite) and [issue_comment](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#issue_comment).

Currently I wrote the program with no dependencies (except for local testing) in order to support the fastest possible runtime of a check enforcer action. If we go with this approach and roll it out, and things look good, I plan to design and add support for publishing a tool instead of building and running from source. At this point dependencies won't have as large of a download overhead and we can introduce them. Primarily I plan to replace my homerolled github client with the primary [github client for go](https://github.com/google/go-github).

For behavior and UX details, as well as instructions to enable, see the README file in this PR at `tools/check-enforcer-actions/README.md`. This tool has functional parity with the current check enforcer.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+name: 'Azure SDK Action'
```
Comment (weshaggard) [inline action.yml:1]: Can we go ahead and setup the structure such that this not in the root but instead a sub-folder so others can move new actions here without needing to update this.

File: action.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    required: true
+  token:
+    description: 'Github event token'
+    required: true
+runs:
+  using: "composite"
+  steps:
+    - run: |
+        echo "Handle webhook event..."
+        cd $GITHUB_ACTION_PATH/src
```
Comment (weshaggard) [inline action.yml:15]: Why not just set this as the working directory?

File: action.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        echo "$GITHUB_PAYLOAD" > ./github_payload.json
+        go run . ./github_payload.json
+      shell: bash
```
Comment (weshaggard) [inline action.yml:18]: Do we expect this to work on windows and mac as well?

File: src/go.mod [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module github.com/azure/azure-sdk-actions
+
+go 1.18
+
+require github.com/stretchr/testify v1.7.1
+
+require (
+	github.com/davecgh/go-spew v1.1.0 // indirect
```
Comment (weshaggard) [inline src/go.mod:8]: How confident are we with taking dependencies on these third-party packages?

File: action.yml (hunk 3 above)
Comment (benbp) [reply action.yml:18]: No, I expect the actions themselves to only run on linux agents, so I'm not worried about the shell type here given it should never be more than few lines (and chose bash for fastest startup). The binary itself I expect to be cross platform though.

File: action.yml (hunk 2 above)
Comment (benbp) [reply action.yml:15]: I was trying to keep too many files out of the root path, but it's probably premature. I'll re-organize.

File: action.yml (hunk 1 above)
Comment (benbp) [reply action.yml:1]: Let me try this out.

File: src/types.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	return &cs
+}
+
+func NewIssueCommentWebhook(payload []byte) *IssueCommentWebhook {
+	var ic IssueCommentWebhook
+	if err := json.Unmarshal(payload, &ic); err != nil {
+		return nil
+	}
+	if ic.Issue.Number == 0 {
```
Comment (weshaggard) [inline src/types.go:186]: Would ic.Issue ever be nil?

File: src/types.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	return &pr
+}
+
+func NewCheckSuiteWebhook(payload []byte) *CheckSuiteWebhook {
+	var cs CheckSuiteWebhook
+	if err := json.Unmarshal(payload, &cs); err != nil {
+		return nil
+	}
+	if cs.CheckSuite.Id == 0 {
```
Comment (weshaggard) [inline src/types.go:175]: Would cs.CheckSuite ever be nil?

File: src/types.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+type StatusBody struct {
+	State       CommitState `json:"state"`
+	Description string      `json:"description"`
+	Context     string      `json:"context"`
+	TargetUrl   string      `json:"target_url"`
+}
+
+type PullRequest struct {
+	Url         string `json:"url"`
```
Comment (weshaggard) [inline src/types.go:45]: I'm not familiar with go but what is the `json:"ur"` telling here? Is that some sort of attribute that the json parser will read to map the value?

File: src/testpayloads/check_suite_response.json [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        "visibility": "public",
+        "pushed_at": "2011-01-26T19:06:43Z",
+        "created_at": "2011-01-26T19:01:12Z",
+        "updated_at": "2011-01-26T19:14:43Z",
+        "permissions": {
+          "admin": false,
+          "push": false,
+          "pull": true
+        },
+        "temp_clone_token": "ABTLWHOULUVAXGTRYU7OC2876QJ2O",
```
Comment (weshaggard) [inline src/testpayloads/check_suite_response.json:158]: What sort of token is this? Is it any sort of secret?

File: src/go.mod (hunk 4 above)
Comment (benbp) [reply src/go.mod:8]: This is only a test dependency and will not be included in current method of action execution (`go run`). When we start publishing a release build of the tool, we can add stuff to make sure these test dependencies aren't included in it.

File: action.yml (hunk 3 above)
Comment (weshaggard) [reply action.yml:18]: I suspect bash will work on windows and mac as well we should be careful about making sure our actions don't end up only working on linux even if that is the primary OS.

File: src/types.go (hunk 5 above)
Comment (benbp) [reply src/types.go:186]: Go uses default zero values for primitive types. So when you instantiate a new struct object, that object and all its struct/string/bool/number fields will be initialized with the default zero value for that type. This basically eliminates this class of issues when not using reference types.

File: action.yml (hunk 3 above)
Comment (benbp) [reply action.yml:18]: What benefit do we get for supporting cross platform actions in the scope of our event handler type actions? I can see cross platform as necessary for our normal test pipelines, but I don't think those belong in this repository.

For local testing with act, it's all containerized so it shouldn't be an issue.

File: src/types.go (hunk 6 above)
Comment (benbp) [reply src/types.go:175]: No, see https://github.com/Azure/azure-sdk-actions/pull/1#discussion_r872804863

File: src/types.go (hunk 7 above)
Comment (benbp) [reply src/types.go:45]: Correct, it's the equivalent of the `JsonProperty` attribute. The go language spec supports arbitrary "struct tags" that packages can consume and use (see [examples](https://zchee.github.io/golang-wiki/Well-known-struct-tags/)).

File: src/testpayloads/check_suite_response.json (hunk 8 above)
Comment (benbp) [reply src/testpayloads/check_suite_response.json:158]: I have no idea, and internet searching is pretty useless. This whole payload body is copied off the github docs [Example Response for check suites](https://docs.github.com/en/rest/checks/suites), so it's not anything sensitive.

File: action.yml (hunk 1 above)
Comment (benbp) [reply action.yml:1]: From what I can tell, external repo actions don't support sub-paths, just an optional git ref ([docs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsstepsuses)). However we could support `with` parameters to the top level action to specify a sub-action, and then we can call `uses` from the main action to reference other `action.yml` files in the repository.

**Pull request merged**